### ISSUE TITLE
audio - video sync problem

### ISSUE THREAD

**Issue by richard42** [mupen64plus/mupen64plus-user-issues#617]
Originally reported on Google Code with ID 618

```
Describe your system:
- Operating System (be specific):
openSUSE 12.3 (Kernel 3.7.10)

- Machine type (32-bit or 64-bit):
64-bit

- Mupen64Plus version:
recent git (25th of may)

- Plugins used:
rsp-hle, glide64mk2, rice


Describe the problem:
in the recent git version i noticed a slight audio delay. this seems to affect all
games, no matter what plugin used.

Steps to reproduce:
1. start the Diddy Kong Racing (E) (M2) (V1.0) [!] rom (CRC: fd73f775 9724755a)
2. right at the title screen enter the options menu
3. when switching between the options this "ting" noise appears, in my case about half
a second too late.

expected result:
the "ting" noise should appear just in the moment when a menu entry is selected


i tried a version from the 22nd of March and everything is fine there.
might be due to a recent change?
```

Reported by `tom.mbrt` on 2014-05-30 16:01:00

**Comment by richard42** [comment]
```
UPDATE: this problem seems to be caused by the new vi timing code, compiling with the
old one and everything works fine
```

Reported by `tom.mbrt` on 2014-06-01 19:16:32

**Comment by richard42** [comment]
OK I have submitted changes to mupen64plus-core and mupen64plus-audio-sdl which will fix this problem.  There is a fundamental design issue in that (for various reasons) we have two pieces of speed limiter code: one in the emulator core (called after each screen refresh) and one in the audio plugin, which runs when new audio data is handed off for output by the running game.  There is currently no API for communicating the audio buffer level to the core, so the two speed limiters can end up fighting with each other.  With the old (problematic) speed limiter code in the core, the emulator would run a tiny bit slow over time and the audio buffer level would stay low, so that audio/video delay would also be low but occasionally you would hear a cutout.  With the new (better) speed limiter code in the core, it runs a tiny bit fast over time, which causes the audio buffer to fill up (and the a/v delay to increase) until it exceeds its target level and the speed limiter in the audio plugin kicks in and inserts delays to avoid overflowing the buffers.  The proper fix is to expand the API to allow the core speed limiter to track the audio plugin's buffer level.  I have implemented a temporary solution by increasing the accuracy of the Core's speed limiter by using floating point variables, and decreasing the default primary audio buffer target level in the audio plugin.
